### DISCUSSION TITLE
OCPP setup on ABB Terra

### DISCUSSION THREAD

**Discussion by a42wg** [lbbrhzn/ocpp#195] [General]
Hi, is TerraConfig the only way to configure the OCPP server. I have not been able to figure out how to get a login via ABB in Denmark

reg Anders

**Comment by lbbrhzn** [comment]
I assume you mean the Terra AC Wallbox?

**Reply by a42wg** [reply]
Yes, I only have access from the ChargerSync app and that don't help much in this case

**Comment by lbbrhzn** [comment]
Looks like you need to preconfigure the URL on the Terra config portal, and then use the app to update your charger.
See https://youtu.be/ixgAI5pPstM.

**Reply by a42wg** [reply]
Yes you are right I found a TerraConfig manual online.
I have tried to write to ABB in Denmark, but they don't deal with private customers. So now I am waiting for response from the guy that sold me the box.
Hoping he can help

reg Anders

**Comment by lbbrhzn** [comment]
You might be able to use the ChargerSync portal instead. It provides similar functionality to the app. Maybe you can enter the OCPP url there.

**Comment by a42wg** [comment]
Thanks, I will give it a try, when I get home

/A

**Reply by Sortberg** [reply]
Hello. Did you get it to work with Home Assistant

Jacob Denmark.

**Reply by anders-westergaard** [reply]
Yes, it is working just fine, but setup was troublesome. ABB have recently deployed new apps and stuff, so I am not sure this is the right procedure any more, but what I did was: Had a friend of mine whom is an authorized electrician, request an installer login from ABB, login to the management portal and define the OCPP settings there. After that I was able to select the OCPP config from the Charge Sync app and that was that. 
Note: I am running the charger on WiFi and when ever the charger looses the connection to HA (one or the other restarts) it can take up to 20 min to reconnect. It may be faster if running via cable. At least in recent firmware's the WiFi has become quite stable, which wasn't the case in the beginning.
Bottom line it is stable now, I don't give it much thoughts and the integration just works

**Reply by Sortberg** [reply]
That's great to hear! I've finally gotten it into Home Assistant too. However, I'm getting a lot of error messages, like: OCPP integration Security event notification received: SettingSystemTime. Are you getting those too? Or have you managed to disable them?

**Reply by a42wg** [reply]
I don't use the charger much at the moment, got a car with one year free charge, but I do get some notifications now and then. I have this automation, which seems to suppress most of them. The "input_boolean.ocpp_notifications" toggles the automation on or off:

alias: OCPP notification
description: ""
mode: parallel
max: 10
triggers:
  - update_type:
      - added
    notification_id: ""
    trigger: persistent_notification
conditions:
  - condition: and
    conditions:
      - condition: state
        entity_id: input_boolean.ocpp_notifications
        state: "off"
      - condition: template
        value_template: "{{ trigger.notification.title | lower == \"ocpp integration\" }}"
actions:
  - delay:
      hours: 0
      minutes: 10
      seconds: 0
      milliseconds: 0
  - data:
      notification_id: "{{ trigger.notification.notification_id }}"
    action: persistent_notification.dismiss

**Comment by stigkildegaard** [comment]
Hi, I am in the same situation as a42wg was, and I am very interested if/how you made it work?

My current plan is to ask my electrician to configure the OCPP server settings on my charger. I would like to know what settings to ask to have configure so that it will work without too much trial and error. I kindly ask that you review if the settings below for the ABB TerraConfig app are correct (or point me to somewhere, where I can read what to set):

Third Party Name: Home assistant Local
Alias:
Type: ws:// (or should it be wss:// ?)
URL: 192.168.1.6 (The local IP address for home assistant on my network)
Protocol Type: OCPP
Password: MyOwnPassword  (Is this the password I use to log onto home assistant, or is it a separate password?)
Repeated PAssword: MyOwnPassword
Certificate Upload: DigiCert Global Root CA
Company: My electricians company?
Instruction:

Thank you in advance!

**Reply by a42wg** [reply]
Hi Stig, I can explain what worked for me.
You need a TerraConfig login and then you need to have the OCCP configuration uploaded to the TerraConfig cloud backend and linked to your account. I am not sure how to upload the configuration a friendly ABB supporter did that for me.
When this is done you login to the TerraConfig App, connect to the charger an choose to configure OCCP.
Now you can download a OCCP config and the one linked to your account will show. 
The attached file show what I sent to the ABB support in order for them to make my config in the cloud service

![Capture](https://user-images.githubusercontent.com/93745277/210892677-b21973ec-7bf4-42bc-a574-25c6f895ccca.JPG)

/Anders

**Reply by AndersV209** [reply]
Hello Anders 

how did you get the TerraConfig login, i have been trying to get one from ABB denmark but they wont work with private customeres.. 

From Anders

**Reply by a42wg** [reply]
Hello, I have had the exact same experience with ABB. If you are an end customer they don't give a sh.. .I'll never buy an ABB product again

I think you need to find an authorized electrician and let him do the initial contact via the contact form on the ABB web site. At least that was what I did. You need your own terra config account and then have your configuration linked to it via the cloud service.

The cloud part I think you can do by yourself at: https://abb.installer.chargedot.com

**Reply by stigkildegaard** [reply]
Hi,

I just wanted to say that my experience has been completely different from yours - in the positive direction. I first asked my electrician for advice. Then I wrote an email to contact.center@dk.abb.com (dk because I am Danish) on behalf of my electrician who installed the charger and asked ABB to put the configuration below into the TerraConfig account of my electrician:

Third party name: LocalHass
Type: ws://
URL: [192.168.1.6:9000](http://192.168.1.6:9000/)
Protocol Type: OCPP
Protocol version: ocpp16j

I send the email in the evening and, honestly, my expectation was that I would have to wait a long time when contacting such a big company and that I would likely just receive some standard copy-pasted nonsense.

So I was pretty blown away when at 09:17 on the following morning I had received three emails from ABB: A new login to the TerraConfig app, a new login to the TerraConfig Web where my server had already been preconfigured and where I could make adjustments to the configuration if needed, and an email explaining what my electrician should do next. With that info it was a 5 min job to log into the new account, connect to the charger, and connect the charger to my HomeAssistant server.

Our electric car has still not arrived so I cannot say how well the OCPP works in practice, but it seems that I can switch the chargers availability on and off and that I can set the maximum charging current.

So, our second charger will most likely also be from ABB :-)

Best regards,
Stig

**Reply by a42wg** [reply]
Well it seems that we have similar experiences as long as you contact ABB via en authorized electrician, but this is where I find ABB's app feature separation plain stupid. You can change the input power on the charger to 32 amps and in theory overload your installation. This can be done from the end-user App, but changing ocpp, which is more or less harmless requires an electrician and a special login. And they don't even bother helping you.
More over I am quite sure that ABB promised OCPP configuration and WiFi access from the Charge Sync App about 1½ ago when I bought the charger, but I haven't found it yet.

I have been using an old version of the OCPP plugin due to incompatibility with the charger and resent changes to how the Charge_Control entity works.

Around new year I did a complete update of the entire HASS system including plugins, but I kept the old OCPP. After that I found out that my charger ended in a boot loop, booting itself every 3 min. I roled all updates back to previous version, just to find out that it still rebooted.
At that point I realized that I also did an update to the charger firmware a week earlier. I contacted ABB for a way to downgrade to the previous version of the firmware - Guess how that went.

Besides the above:

The charger do not like if charge_control is switched too often within a charge cycle, doing so may very well end in a hard reset to make things working again. I have solar panels and try to optimize charging so I send as little power as possible back into the grid and this involve toggling the charger on and off a lot.  

Making the charger connect in Home Assistant is a pain, that often involves rebooting the charger, reloading the plugin rebooting HASS or even reinstalling the plugin.

These observations may not only be the charger, but it certainly don't add to the overall experience. So I am actually considering getting another charger, because I simply don't want to waste more time on this.

If your only experience is that you are able to switch charger availability: You may be in for a surprise.

NB: I ended updating the entire HASS system once again, this time including the OCPP plugin, I then disabled unused entities and ended with a working system - for now

**Reply by stigkildegaard** [reply]
Hi a42wg,

Oh dear, it really does sound like you have been having a lot of trouble! I do not have the benefit of your experience, as I am quite new to Home Assistant and because our EV has not arrived yet. I was just happy that the charger seemed to connect to Home Assistant without much effort.

I agree that from our point of view as end users, it is quite puzzling that every single user needs to involve an electrician and ABB Support just to enable and configure OCPP. But to me it seems likely that there either is, or has previously been, some business reason why ABB made it that way. I assume that the pool of private customers who want to configure OCPP themselves is still so limited that ABB sees more sense in using their resources to develop other stuff. Like you, I hope that they get around to making it easier for end users so that we do not need to involve electricians and ABB support to configure OCPP. But until then I will remain happy that at least ABB support appears to be very fast to support electricians and, thereby, also us. I am sure that they have engineers that want to make the upgrade. If they are anything like the engineers I know, they want to be on the team that has the coolest products that customers love.

It does sound like you have a lot of trouble with connectivity. May I ask if you use a wired connection or a WiFi connection to the charger? Mine is on a wired connection and I have not yet had connectivity issues. But, then again, my experience is very limited and, since I do not have an EV yet, I do not have experience with how the charging actually works.

I would like to ask you why it is necessary to switch the charge control on and off? Off course, I agree completely with you that toggling it should not cause connectivity issues. But my current understanding is so limited that I do not yet understand why it is necessary? My plan was to control the charging only by controlling the charger_maximum_current. I have a simple control algorithm going that has been setting the maximum current every 5 seconds for some days and have not yet experienced any hiccups with that. But then again, no EV yet, so I do not know if it actually works for controlling the charging. I would greatly appreciate, if you could briefly explain why switching the charge control is necessary or perhaps give me a few key words to search for so that I may read about the subject to learn more.

Thanks in advance for any feedback!

**Reply by a42wg** [reply]
Think I forget to mention that the first supporter I got in contact with answered quickly, asked for the setup, which I sent him and then forgot all about the case. I was a bit reluctant to put pressure on him because I feared they would refuse to help. After a week I called him and he had forgot an would take action immediately. After another week I called again and he sent me a generic test login, which was useless for other that preview. Called once again and lucky me, he was on vacation and the substitute made things happen in about 5 min.
So bottom line I am not impressed at all.

I am connected via WiFi, but that does not seems to be the problem WiFi connects right away, but the charger do not show up in Home Assistant and all the entities are "Unavailable". That said after installing the new firmware and HA plugin things seems to way more stable. Previous the charger could drop out for hours without any obvious reason. Now it stay connected.

I have solar panels and production is calculated by the hour. As I want to use my production as effectively as possible I strive to use the production myself within the production hour, instead of sending it to the grid. So when ever I have a negative consumption i feed power into my car, my hot water reservoir and my heating/cooling pump, in that prioritized order. But because I also want to be able to charge my car elsewhere, I have configured the car to charge when ever it is plugged in and when at home I control the charging cycles by turning Charge_Control on and off. This may not be the only way to control charging, but it have been rather complex to get it running, so I don't want to mess too much with it.

**Comment by Kooees** [comment]
Hey

I'm having trouble getting Abb Terra added to home assistant. I have a credentials for the config portal and made profiileja following the instructions. In the config app, I can add an ocpp server, but after rebooting the charger, it does not get an internet connection. I have tried ocpp profiileja with the ip address and homeassistant.local and alko with wifi and lan connection. In the Terra config portal, the ocpp profile also has a password field, does it have to be something?

Any tips on where to start looking for the problem?
thanks in advance for your help

**Reply by stigkildegaard** [reply]
Hi Kooees,

I am unsure what you mean by "made profiileja following the instructions", so I will try to explain how the configuration looks for me where the charger and the Home Assistant server are on the same LAN in the hope that it helps you.

My OCPP server configuration in the ABB TerraConfig portal is shown in the attached picture. You need to adapt the IP address, so that it matches the IP of your home assistant server.
![ABB OCPP Server config upload](https://user-images.githubusercontent.com/120266686/231821970-118c06c9-c489-4dc7-b41f-3a17cb29eb1c.png)

In the TerraConfig app you first need to connect to your charger.  Then under "Configuration" choose "OCPP Server" and the above profile should appear as a "Locally Cached Server", so that you can select it and press next.
![Screenshot_20230413_181807_TerraConfig](https://user-images.githubusercontent.com/120266686/231823445-893fab32-f00b-48d7-b771-687765fde02e.jpg)

On the next screen press configure:
![Screenshot_20230413_183857_TerraConfig](https://user-images.githubusercontent.com/120266686/231842197-6cc0af3f-b931-4ee8-888a-641eb87e4362.jpg)

After a few seconds you should get three blue check marks like shown here:
 ![Screenshot_20230413_183917_TerraConfig](https://user-images.githubusercontent.com/120266686/231842388-1852b697-3ef2-4ac4-a6a3-aaf5aedcfa18.jpg)

Then reboot the charger and reboot home assistant, and they should connect. For me the connection has worked flawlessly for controlling the charging of an EV since it arrived about two months ago.

Best regards,
Stig

**Reply by Kooees** [reply]
Thank you Stig! 

It’s working now!

**Comment by arn1s-m** [comment]
If you stumbled on this thread by searching "How to connect ABB Terra to Home Assistant over OCPP (or to ocpp in general)" then I'll post latest "how to".
1. You will need TerraConfig app (not the legacy one, there's 2 options atm)
2. Configure charger using it, go to "OCPP Server" settings
3. Enable "External server" and configure "Custom server - available on your device only"
4. Server URL type in "ws://IP_address or hostname:9000/" (wss for secure) of the HA instance
5. Save, upload the configuration to charger.
6. Install OCPP via HACS in Home Assistant.
7. Set up the "central" OCPP as per guide
8. Restart charger/HA and option to add the charger will appear by itself

**Reply by KESchnitzler** [reply]
Hi,
based on arn1s-m tutorial i tried to install my terra ac wallbox. after restarting both (HA & WB) ocpp integration recovered 2 devices (central & WB) together with 52 entities - but all these entities are marked unknown? is there any advice to solve this problem??
regs K-E

**Reply by arn1s-m** [reply]
@KESchnitzler - Does the Terra connect to Wifi (or LAN), see second light from bottom blinking or being solid green?. It sometime takes ~20+ sec to connect to Wifi, then some to see it active in HA.
Maybe you see some errors in HA? OCPP integration (last version) has an issue to add new devices, so use version prior that.

Also, adding to HA you need to select which entities you need (see OCPP documentation for Terra)

**Reply by KESchnitzler** [reply]
hi,
thanks for you quick advises. the WB is connected via LAN, the inet led is solid green. use the actual versin v0.8.4. will try to downgrade maybe tomorrow evening. the logfiles of the 2 ocpp devices (central/WB) are empty. HA log see image.
![OCPP Error HA Core Log](https://github.com/user-attachments/assets/a2e78a59-dd1c-488e-9298-dc5ce52dc7ce)
regs K-E

**Reply by arn1s-m** [reply]
Use the 0.8.3, there's bug report opened by others that new device setup isn't possible with it.
At the end you'll have 2 entities in HA - "central" as server in a sense and the actual Terra. I think with version downgrade you will be fine

**Comment by KESchnitzler** [comment]
surprise, this morning after the actual HA updates, the ocpp front end seems to be active wo installing the v0.8.3. i'll add the infos that are now available
![diagnose](https://github.com/user-attachments/assets/fdbfb2fa-e468-44a4-b43d-09c5046159e6)
![Konfig](https://github.com/user-attachments/assets/5dcca4e8-8a75-4114-97a1-59c9f5b4cd59)
![sensoren](https://github.com/user-attachments/assets/ee2f17b2-2030-4e36-93f3-b0d94233ecb3)
![Steuer](https://github.com/user-attachments/assets/07447c2b-4060-450a-a8e0-3d52de4ce171)
so my problem is now how to setup the charging with variable current/power, change from 1 to 3 phase charging to achieve surplus charging for the wb??? for the moment I'm totally lost. ich i see it right, i only can switch charging on/off there seems to be no way to control deeper features of the system and the 2 avail sensors show the total amount of energy that has been charged in my ev... maybe someone can show me the way
K-E

**Reply by arn1s-m** [reply]
Hi @KESchnitzler - you can only adjust power (Amps) by sliding the "Maximum Current". "Charge Control" switches charging on/off. In my case it's 5-16A (with <5A it switched automatically charging off for some reason). I'm not aware of such function like 1 or 3 phase charging since this depends on how charger is wired. Is such option in original ABB app? I haven't used it, so I'm not sure.

To achieve surplus charging have a look at https://github.com/lbbrhzn/ocpp/blob/main/docs/Charge_automation.md
In my setup I have Shelly 3EM measuring import/export on each phase and automation looks like this:
```
alias: Terra Solar charging
description: ""
triggers:
  - seconds: /15
    trigger: time_pattern
conditions:
  - condition: template
    value_template: >
      {{ (states('number.charge_limit') | int - states('sensor.battery') | int)
      > 1 }}
  - condition: state
    entity_id: binary_sensor.charger
    state: "on"
  - condition: device
    type: is_on
    device_id: 4cbd26a068cef9f5e313ca9f358ffce2
    entity_id: 8e9998eba7a7a14f0356f586334c6717
    domain: switch
actions:
  - variables:
      phase_a: "{{ states('sensor.shelly_3em_channel_a_power') | float(0) }}"
      phase_b: "{{ states('sensor.shelly_3em_channel_b_power') | float(0) }}"
      phase_c: "{{ states('sensor.shelly_3em_channel_c_power') | float(0) }}"
      current_setting: "{{ states('number.terra_maximum_current') | float(0) }}"
      max_current: 16
      importing_detected: |
        {{ phase_a > 0 or phase_b > 0 or phase_c > 0 }}
      min_export: >
        {% set exports = [phase_a, phase_b, phase_c] %} {% set export_only =
        exports | select('lt', 0) | list %} {% if export_only | length == 3 %}
          {{ export_only | map('abs') | min }}
        {% else %}
          0
        {% endif %}
      export_amps: "{{ (min_export / 230) | round(0, 'floor') }}"
      target_amps: |
        {% if importing_detected %}
          {{ (current_setting - 1) if current_setting > 1 else 0 }}
        {% else %}
          {% set target = current_setting + export_amps %}
          {{ (target if target < max_current else max_current) | round(0, 'floor') }}
        {% endif %}
      clamped_amps: |
        {% if target_amps < 1 %}
          0
        {% else %}
          {{ target_amps }}
        {% endif %}
      tx_id: "{{ states('sensor.terra_transaction_id') | int(0) }}"
  - data:
      conn_id: 1
      custom_profile: |
        {
          "transactionId": {{ tx_id }},
          "chargingProfileId": 1,
          "stackLevel": 0,
          "chargingProfilePurpose": "TxProfile",
          "chargingProfileKind": "Relative",
          "chargingSchedule": {
            "chargingRateUnit": "A",
            "chargingSchedulePeriod": [
              {
                "startPeriod": 0,
                "limit": {{ clamped_amps }}
              }
            ]
          }
        }
    action: ocpp.set_charge_rate
mode: single
```

In short - every 15 sec it checks each phase import/export state and adjusts the A (charge value) to not import anything. Charge until car battery set limit reached. Use ChatGPT or other tools to modify, adjust automation if you can't figure it out

**Reply by KESchnitzler** [reply]
@arn1s-m thanks!! what language has been used, python??
K-E

**Reply by arn1s-m** [reply]
@KESchnitzler Automation is using YAML (https://www.home-assistant.io/docs/automation/yaml/) but variables in it are in Jinja (https://www.home-assistant.io/docs/configuration/templating/)

**Reply by KESchnitzler** [reply]
uuf, thats nothing for rookies :-(

**Reply by arn1s-m** [reply]
@KESchnitzler - maybe not, but for that reason I suggested to use ChatGPT (or any other) AI tool that will generate automation code instead for you

**Reply by arn1s-m** [reply]
Short update. After further reviewing of docs I came to conclusion that adjusting and using terra_maximum_current sensor is not goot. So, to fully automate need to use 3 things:
1. Create new slider (in my case with 6-16 values, for some reason setting it to 5 switches charging off)
2. Create automation that will adjust the charging Amps when above slider is moved:
```
alias: Adjust Terra Charge Amps
description: ""
triggers:
  - trigger: state
    entity_id:
      - switch.terra_charge_control
    to: "on"
    id: Switch On
    alias: Switch ON
  - trigger: state
    entity_id:
      - input_number.terraampcontrol
    alias: Adjust Amps
conditions:
  - condition: state
    entity_id: sensor.terra_status_connector
    state:
      - Preparing
      - Charging
      - Finishing
      - SuspendedEV
actions:
  - alias: If triggered by Switch On, delay for startup
    if:
      - condition: trigger
        id:
          - Switch On
    then:
      - delay:
          hours: 0
          minutes: 0
          seconds: 20
          milliseconds: 0
  - alias: If charger is on, set the charge rate
    if:
      - condition: state
        entity_id: switch.terra_charge_control
        state: "on"
    then:
      - action: ocpp.set_charge_rate
        data:
          custom_profile: |
            {
              "transactionId": {{ states('sensor.terra_transaction_id') | int }},
              "chargingProfileId": 1,
              "stackLevel": 0,
              "chargingProfilePurpose": "TxProfile",
              "chargingProfileKind": "Relative",
              "chargingSchedule": {
                "chargingRateUnit": "A",
                "chargingSchedulePeriod": [
                  {"startPeriod": 0, "limit": {{ min(states('input_number.terraampcontrol') | int,states('number.terra_maximum_current') | int) }}}  
                ]
              }
            }
          conn_id: 1
      - if:
          - condition: numeric_state
            entity_id: input_number.terraampcontrol
            above: number.terra_maximum_current
        then:
          - action: input_number.set_value
            metadata: {}
            data:
              value: "{{ states('number.terra_maximum_current') }}"
            target:
              entity_id: input_number.terraampcontrol
mode: queued
max: 5
trace:
  stored_traces: 25
```
3. Create the automation that will use excess solar to charge
```
alias: Terra Solar charging
description: ""
triggers:
  - seconds: /20
    trigger: time_pattern
conditions:
  - condition: template
    value_template: >
      {{ (states('number.charge_limit') | int - states('sensor.battery') | int)
      > 1 }}
  - condition: state
    entity_id: binary_sensor.charger
    state: "on"
  - condition: device
    type: is_on
    device_id: 4cbd26a068cef9f5e313ca9f358ffce2
    entity_id: 8e9998eba7a7a14f0356f586334c6717
    domain: switch
actions:
  - variables:
      phase_a: "{{ states('sensor.shelly_3em_channel_a_power') | float(0) }}"
      phase_b: "{{ states('sensor.shelly_3em_channel_b_power') | float(0) }}"
      phase_c: "{{ states('sensor.shelly_3em_channel_c_power') | float(0) }}"
      current_setting: "{{ states('input_number.terraampcontrol') | float(0) }}"
      max_current: 16
      importing_detected: |
        {{ phase_a > 0 or phase_b > 0 or phase_c > 0 }}
      min_export: >
        {% set exports = [phase_a, phase_b, phase_c] %} {% set export_only =
        exports | select('lt', 0) | list %} {% if export_only | length == 3 %}
          {{ export_only | map('abs') | min }}
        {% else %}
          0
        {% endif %}
      export_amps: "{{ (min_export / 230) | round(0, 'floor') }}"
      target_amps: |
        {% if importing_detected %}
          {{ (current_setting - 1) if current_setting > 1 else 0 }}
        {% else %}
          {% set target = current_setting + export_amps %}
          {{ (target if target < max_current else max_current) | round(0, 'floor') }}
        {% endif %}
      clamped_amps: |
        {% if target_amps < 1 %}
          0
        {% else %}
          {{ target_amps }}
        {% endif %}
      tx_id: "{{ states('sensor.terra_transaction_id') | int(0) }}"
  - data:
      conn_id: 1
      custom_profile: |
        {
          "transactionId": {{ tx_id }},
          "chargingProfileId": 1,
          "stackLevel": 0,
          "chargingProfilePurpose": "TxProfile",
          "chargingProfileKind": "Relative",
          "chargingSchedule": {
            "chargingRateUnit": "A",
            "chargingSchedulePeriod": [
              {
                "startPeriod": 0,
                "limit": {{ clamped_amps }}
              }
            ]
          }
        }
    action: ocpp.set_charge_rate
    enabled: false
  - action: input_number.set_value
    metadata: {}
    data:
      value: "{{ clamped_amps }}"
    target:
      entity_id: input_number.terraampcontrol
mode: single
```
These should extend the lifespan of the charger

**Reply by cver-me** [reply]
Thank you, @arn1s-m . If I use the new slider to change the amperage, what value is then set for the maximum_current entity?

**Reply by arn1s-m** [reply]
@cver-me - set that to the maximum what your charger/house supports i.e. 16A (3 phase setup should be then 11kw charger). No need to touch that value after setting it at all

**Reply by nixmeer** [reply]
@arn1s-m I try to modify the charge limit as you described above. 
```
action: ocpp.set_charge_rate
data:
  custom_profile: |
    {
      "transactionId": {{ states('sensor.terraac_transaction_id') | int }},
      "chargingProfileId": 1,
      "stackLevel": 0,
      "chargingProfilePurpose": "TxProfile",
      "chargingProfileKind": "Relative",
      "chargingSchedule": {
        "chargingRateUnit": "A",
        "chargingSchedulePeriod": [
          {"startPeriod": 0, "limit": {{ 16 | int }} }  
        ]
      }
    }
  conn_id: 1
  ```
  The logs show:
  ```
2026-05-01 20:38:46.289 INFO (MainThread) [ocpp] TACW1143921G0370: send [2,"3759345c-eab7-4425-a06b-d9f45867ac99","SetChargingProfile",{"connectorId":1,"csChargingProfiles":{"transactionId":1777656812,"chargingProfileId":1,"stackLevel":0,"chargingProfilePurpose":"TxProfile","chargingProfileKind":"Relative","chargingSchedule":{"chargingRateUnit":"A","chargingSchedulePeriod":[{"startPeriod":0,"limit":16}]}}}]
2026-05-01 20:38:46.293 INFO (MainThread) [ocpp] TACW1143921G0370: receive message [3, "3759345c-eab7-4425-a06b-d9f45867ac99", {"status": "Accepted"}]
 ```
  I also tried 0 and 1 for `chargingProfileId` and 0, 1 and 2 for `stackLevel`. `stackLevel` of 3 was rejected.
  
Did you get your code to work? If so, what exact model of the Terra AC do you use? Is there anything different in our settings?

**Reply by arn1s-m** [reply]
> @arn1s-m I try to modify the charge limit as you described above.
> 
> ```
> action: ocpp.set_charge_rate
> data:
>   custom_profile: |
>     {
>       "transactionId": {{ states('sensor.terraac_transaction_id') | int }},
>       "chargingProfileId": 1,
>       "stackLevel": 0,
>       "chargingProfilePurpose": "TxProfile",
>       "chargingProfileKind": "Relative",
>       "chargingSchedule": {
>         "chargingRateUnit": "A",
>         "chargingSchedulePeriod": [
>           {"startPeriod": 0, "limit": {{ 16 | int }} }  
>         ]
>       }
>     }
>   conn_id: 1
> ```
> 
> The logs show:
> 
> ```
> 2026-05-01 20:38:46.289 INFO (MainThread) [ocpp] TACW1143921G0370: send [2,"3759345c-eab7-4425-a06b-d9f45867ac99","SetChargingProfile",{"connectorId":1,"csChargingProfiles":{"transactionId":1777656812,"chargingProfileId":1,"stackLevel":0,"chargingProfilePurpose":"TxProfile","chargingProfileKind":"Relative","chargingSchedule":{"chargingRateUnit":"A","chargingSchedulePeriod":[{"startPeriod":0,"limit":16}]}}}]
> 2026-05-01 20:38:46.293 INFO (MainThread) [ocpp] TACW1143921G0370: receive message [3, "3759345c-eab7-4425-a06b-d9f45867ac99", {"status": "Accepted"}]
> ```
> 
> I also tried 0 and 1 for `chargingProfileId` and 0, 1 and 2 for `stackLevel`. `stackLevel` of 3 was rejected.
> 
> Did you get your code to work? If so, what exact model of the Terra AC do you use? Is there anything different in our settings?

I've since tweaked the code a bit. The slider has remained as is.
2. adjust the charging Amps when above slider is moved
```
alias: Adjust Terra Charge Amps
description: ""
triggers:
  - alias: Switch ON
    trigger: state
    entity_id:
      - switch.charge_control
    to:
      - "on"
    id: Switch On
  - trigger: state
    entity_id:
      - input_number.terraampcontrol
    alias: Adjust Amps
conditions:
  - condition: state
    entity_id: sensor.terra_status_connector
    state:
      - Preparing
      - Charging
      - Finishing
      - SuspendedEVSE
      - SuspendedEV
actions:
  - alias: If triggered by Switch On, delay for startup
    if:
      - condition: trigger
        id:
          - Switch On
    then:
      - delay:
          hours: 0
          minutes: 0
          seconds: 20
          milliseconds: 0
  - alias: If charger is on, set the charge rate
    if:
      - condition: state
        entity_id: switch.charge_control
        state:
          - "on"
    then:
      - action: ocpp.set_charge_rate
        data:
          custom_profile: |
            {
              "transactionId": {{ states('sensor.terra_transaction_id') | int }},
              "chargingProfileId": 1,
              "stackLevel": 0,
              "chargingProfilePurpose": "TxProfile",
              "chargingProfileKind": "Relative",
              "chargingSchedule": {
                "chargingRateUnit": "A",
                "chargingSchedulePeriod": [
                  {"startPeriod": 0, "limit": {% set raw = min(states('input_number.terraampcontrol') | int, states('number.terra_maximum_current') | int) %}{{ 0 if raw < 6 else raw }}}
                ]
              }
            }
          conn_id: 1
      - if:
          - condition: numeric_state
            entity_id: input_number.terraampcontrol
            above: number.terra_maximum_current
        then:
          - action: input_number.set_value
            metadata: {}
            data:
              value: "{{ states('number.terra_maximum_current') }}"
            target:
              entity_id: input_number.terraampcontrol
mode: queued
max: 5
trace:
  stored_traces: 25
```

3.  automation that will use excess solar to charge
```alias: Terra Solar charging
description: Calculates optimal charge amps from per-phase solar excess
triggers:
  - seconds: /20
    trigger: time_pattern
conditions:
  - condition: template
    value_template: >
      {{ (states('number.charge_limit') | int - states('sensor.battery') | int)
      >= 1 }}
  - condition: state
    entity_id: binary_sensor.charger
    state: "on"
actions:
  - variables:
      phase_a: "{{ states('sensor.shelly_3em_channel_a_power') | float(0) }}"
      phase_b: "{{ states('sensor.shelly_3em_channel_b_power') | float(0) }}"
      phase_c: "{{ states('sensor.shelly_3em_channel_c_power') | float(0) }}"
      current_setting: "{{ states('input_number.terraampcontrol') | float(0) }}"
      max_current: 16
      worst_phase: "{{ [phase_a, phase_b, phase_c] | max }}"
      target_amps: |
        {% if worst_phase | float > 100 %}
          {% set reduce_by = ((worst_phase | float / 240) | round(0, 'ceil')) | int %}
          {{ [(current_setting | int - reduce_by), 5] | max }}
        {% elif worst_phase | float < -200 %}
          {% set headroom = ((worst_phase | float | abs) / 240) | round(0, 'floor') | int %}
          {{ [(current_setting | int + headroom), max_current] | min }}
        {% else %}
          {{ current_setting | int }}
        {% endif %}
  - condition: template
    value_template: "{{ target_amps | int != current_setting | int }}"
  - action: input_number.set_value
    data:
      value: "{{ target_amps }}"
    target:
      entity_id: input_number.terraampcontrol
mode: single
```

Biggest change was related to charger state recovery i.e. in original script when it went <6A, it switched off and didn't switch back on. Now code handles that nicely. Works fine for me with solar monitoring. Occasionally it can happen that amps don't update if there's some communication error and the requests get queued but that happens so rare that I haven't bothered to make separate handler.
I have tac-w11-g5-r-0
